**Problem 1: Word counts**

*Part A. Characters in Little Women*

*How many times are each of the following characters mentioned by name in the text of Little Women?:* 

*Jo, Beth, Meg, Amy*

*Use the text available at https://raw.githubusercontent.com/gwsb-istm-6212-fall-2016/syllabus-andschedule/
master/projects/project-01/women.txt for this part.*

Step 1: Retrieve text of Little Women

In [1]:
!wget https://raw.githubusercontent.com/gwsb-istm-6212-fall-2016/syllabus-and-schedule/master/projects/project-01/women.txt

--2016-09-23 01:59:06--  https://raw.githubusercontent.com/gwsb-istm-6212-fall-2016/syllabus-and-schedule/master/projects/project-01/women.txt
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 151.101.20.133
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|151.101.20.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1053440 (1.0M) [text/plain]
Saving to: ‘women.txt.1’

women.txt.1         100%[=====================>]   1.00M  --.-KB/s   in 0.05s  

2016-09-23 01:59:06 (22.0 MB/s) - ‘women.txt.1’ saved [1053440/1053440]



Step 2: Produce counts of the names 'Jo', 'Beth', 'Meg', 'Amy'. Consider matches regardless of upper/lower case.

In [3]:
!grep -oE '\w{{2,}}' women.txt |\
    tr '[:upper:]' '[:lower:]' |\
    grep -w 'jo\|beth\|meg\|amy' |\
    sort |\
    uniq -c |\
    sort -rn

   1362 jo
    686 meg
    652 amy
    467 beth


The 'grep -w' command is used to limit results to only the exact matches of each name. Without using this, the grep was returning hits for any words that contained the target strings. After applying this function, I sorted by name and retrieved the unique counts. Finally, I presented these counts in descending order, which I have reproduced below:

Jo: 1,362  
Meg: 686  
Amy: 652  
Beth: 467

*Part B. Juliet and Romeo in Romeo and Juliet*

*How many times do each of the characters Juliet and Romeo have speaking lines in Romeo and Juliet? Keep
in mind that this is the text of a play.
Use the text available at https://raw.githubusercontent.com/gwsb-istm-6212-fall-2016/syllabus-andschedule/
master/projects/project-01/romeo.txt for this part.*

Step 1: Retrieve text of Romeo and Juliet

In [36]:
!wget https://raw.githubusercontent.com/gwsb-istm-6212-fall-2016/syllabus-and-schedule/master/projects/project-01/romeo.txt

--2016-09-16 02:26:31--  https://raw.githubusercontent.com/gwsb-istm-6212-fall-2016/syllabus-and-schedule/master/projects/project-01/romeo.txt
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 151.101.20.133
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|151.101.20.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 178983 (175K) [text/plain]
Saving to: ‘romeo.txt’

romeo.txt           100%[=====================>] 174.79K  --.-KB/s   in 0.02s  

2016-09-16 02:26:32 (7.50 MB/s) - ‘romeo.txt’ saved [178983/178983]



Step 2: Calculate number of times Juliet and Romeo have speaking lines

In [41]:
!grep -oE '\w{{2,}}' romeo.txt |\
    grep -w 'Rom\|Jul' |\
    sort |\
    uniq -c

    117 Jul
    163 Rom


To find the number of speaking lines, I used 'Rom' and 'Jul' as the search terms. I used these because in reviewing the text, each time either of these characters had a speaking line, it was prefixed by either 'Rom.' or 'Jul.'. Again, I used the '-w' command to limit it to exact matches. The exact number of times each character spoke is reproduced below:

Juliet: 117  
Romeo: 163

**Problem 2: Capital Bikeshare **

*Part A: Which 10 Capital Bikeshare stations were the most popular departing stations in Q1 2016? Which 10 were the most popular destination stations in Q1 2016?*

Step 1: Retrieve data

In [21]:
!wget https://raw.githubusercontent.com/gwsb-istm-6212-fall-2016/syllabus-and-schedule/master/projects/project-01/2016q1.csv.zip

--2016-09-18 02:19:12--  https://raw.githubusercontent.com/gwsb-istm-6212-fall-2016/syllabus-and-schedule/master/projects/project-01/2016q1.csv.zip
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 151.101.20.133
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|151.101.20.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 10643003 (10M) [application/octet-stream]
Saving to: ‘2016q1.csv.zip’

2016q1.csv.zip      100%[=====================>]  10.15M  16.3MB/s   in 0.6s   

2016-09-18 02:19:13 (16.3 MB/s) - ‘2016q1.csv.zip’ saved [10643003/10643003]



Step 2: Unzip file downloaded in Step 1

In [22]:
!unzip 2016q1.csv.zip

Archive:  2016q1.csv.zip
  inflating: 2016q1.csv              


Step 3: View columns in csv file

In [23]:
!csvcut -n 2016q1.csv

  1: Duration (ms)
  2: Start date
  3: End date
  4: Start station number
  5: Start station
  6: End station number
  7: End station
  8: Bike number
  9: Member Type


Step 4: View sample data for fields: Start date, End date, Start Station, End station

In [25]:
!csvcut -c2,3,5,7,8 2016q1.csv | head -10 | csvlook

|------------------+-----------------+--------------------------------+-------------------------------+--------------|
|  Start date      | End date        | Start station                  | End station                   | Bike number  |
|------------------+-----------------+--------------------------------+-------------------------------+--------------|
|  3/31/2016 23:59 | 4/1/2016 0:04   | 11th & S St NW                 | 1st & Rhode Island Ave NW     | W00022       |
|  3/31/2016 23:59 | 4/1/2016 0:08   | New Hampshire Ave & 24th St NW | 18th St & Wyoming Ave NW      | W01294       |
|  3/31/2016 23:59 | 4/1/2016 0:08   | 14th & V St NW                 | 18th & M St NW                | W01416       |
|  3/31/2016 23:57 | 4/1/2016 0:09   | 34th St & Wisconsin Ave NW     | 17th & Corcoran St NW         | W01090       |
|  3/31/2016 23:57 | 3/31/2016 23:59 | 23rd & Crystal Dr              | 27th & Crystal Dr             | W21934       |
|  3/31/2016 23:57 | 4/1/2016 0:13   | 11th & M 

Step 5: Confirm that all End dates occurred in Q1 2016

In [26]:
!csvcut -c3 2016q1.csv | csvstat

  1. End
	<class 'datetime.date'>
	Nulls: False
	Min: 2016-01-01
	Max: 2016-04-01
	Unique values: 88
	5 most frequent values:
		2016-03-26:	14102
		2016-03-16:	13355
		2016-03-24:	13341
		2016-03-23:	12332
		2016-03-25:	12224
  2. date
	<class 'datetime.time'>
	Nulls: False
	Min: 00:00:00
	Max: 23:59:00
	Unique values: 1440
	5 most frequent values:
		17:53:00:	1193
		17:54:00:	1187
		17:44:00:	1182
		17:43:00:	1177
		17:56:00:	1176

Row count: 552399


Step 6: Save sorted list of [start stations] as csv file. Due to the file size and processing time, I've included this intermediate step of saving the preliminary output to a csv file -- which I repeat for similar exercises below.

In [39]:
!csvcut -c5 2016q1.csv |\
    sort > output1.csv

Step 7: Return top 10 most popular departing stations

In [42]:
!cat output1.csv |\
    uniq -c |\
    sort -rn |\
    head -10

  13120 Columbus Circle / Union Station
   9560 Massachusetts Ave & Dupont Circle NW
   9388 Lincoln Memorial
   8138 Jefferson Dr & 14th St SW
   7479 Thomas Circle
   7401 15th & P St NW
   6568 14th & V St NW
   6491 New Hampshire Ave & T St NW
   5649 Eastern Market Metro / Pennsylvania Ave & 7th St SE
   5514 17th & Corcoran St NW


Step 8: Save sorted list of [end stations] as csv file

In [43]:
!csvcut -c7 2016q1.csv |\
    sort > output2.csv

Step 9: Return top 10 most popular destination stations

In [44]:
!cat output2.csv |\
    uniq -c |\
    sort -rn |\
    head -10

  13880 Columbus Circle / Union Station
  11183 Massachusetts Ave & Dupont Circle NW
   9419 Lincoln Memorial
   8975 Jefferson Dr & 14th St SW
   8092 15th & P St NW
   7267 14th & V St NW
   6997 Thomas Circle
   6245 New Hampshire Ave & T St NW
   5761 5th & K St NW
   5651 17th & Corcoran St NW


*Part B: For the most popular departure station, which 10 bikes were used most in trips departing from there? Which
10 bikes were used most in trips ending at the most popular destination station?*

Step 1: Save sorted list of bikes associated with most popular departure station 'Columbus Circle / Union Station' as csv file

In [46]:
!csvcut -c5,8 2016q1.csv |\
    csvgrep -c1 -m 'Columbus Circle / Union Station' |\
    csvcut -c2 |\
    sort > output3.csv
   

Step 2: Return 10 most popular bike IDs associated with departure station 'Columbus Circle / Union Station' 

In [52]:
!cat output3.csv |\
    uniq -c |\
    sort -rn |\
    head -10

     17 W22227
     16 W21867
     16 W21641
     16 W21538
     16 W21239
     16 W20540
     16 W00714
     15 W22080
     15 W21450
     15 W21076
sort: write failed: standard output: Broken pipe
sort: write error


Step 3: Save sorted list of bikes associated with most popular destination station 'Columbus Circle / Union Station' as csv file

In [48]:
!csvcut -c7,8 2016q1.csv |\
    csvgrep -c1 -m 'Columbus Circle / Union Station' |\
    csvcut -c2 |\
    sort > output4.csv

Step 4: Return 10 most popular bike IDs associated with destination station 'Columbus Circle / Union Station' 

In [53]:
!cat output4.csv |\
    uniq -c |\
    sort -rn |\
    head -10

     18 W00485
     17 W22227
     16 W22099
     16 W22080
     16 W21239
     16 W21076
     16 W20425
     16 W00714
     15 W21997
     15 W21867
sort: write failed: standard output: Broken pipe
sort: write error


**Problem 3: Filters**

*Part A: Demonstrate a pipeline that performs a count of the top ten unique words in Little Women. This may be
exactly the same pipeline we have used before.*      
  
*Write a Python filter than replaces grep -oE '\w{2,}' to split lines of text into one word per line, and
write an additional Python filter to replace tr '[:upper:]' '[:lower:]' to transform text into lower case.*      

*With your two new filters, repeat the original pipeline, and substitute your new filters as appropriate. You
should obtain the same results.*

Step 1: Top 10 words in Little Women using same pipeline as used before. The only change was changing the value in brackets from 2 to 1, so that it would include 1 character words such as "A" and "I" in the counts, which is what was returned by the Python code that follows.

In [26]:
!grep -oE '\w{{1,}}' women.txt |\
    tr '[:upper:]' '[:lower:]' |\
    sort |\
    uniq -c |\
    sort -rn |\
    head -10

   8155 and
   7689 the
   5152 to
   4531 a
   4003 i
   3523 of
   3245 her
   2774 it
   2503 in
   2447 you
sort: write failed: standard output: Broken pipe
sort: write error


Step 2: Produce same analysis as above except using Python filter. I did a head and a tail at the end because the counts originally include blanks as words, which had the largest volume. Doing the head then tail excludes that first record.

In [2]:
!chmod +x split.py

In [3]:
!chmod +x lower.py

In [14]:
!cat women.txt |\
    ./split.py |\
    ./lower.py |\
    sort |\
    uniq -c |\
    sort -rn |\
    head -11 |\
    tail -10

sort: write failed: standard output: Broken pipe
   8098 and
   7681 the
   5148 to
sort: write error
   4509 a
   3518 of
   3245 her
   3144 i
   2500 in
   2470 it
   2278 she


I was unable to get my Python counts to match exactly my grep counts. I believe this is due to differences in how the grep is splitting words opposed to how the Python split function operates. Specifically, I believe that certain special characters that appear next to words are being included in the text of the word. I've attempted to account for this by using a strip command in my split filter which has helped, but gaps still remain. Given more time I would attempt to find a different Python package to be used for splitting text that uses the same logic as the grep command.

*Part B: Write a Python filter that removes at least ten common words of English text, commonly known as “stop
words”. Sources of English stop word lists are readily available online, or you may generate your own list
from the text.
Add your stop word filter to a word count pipeline and show the top 25 words in Little Women with stop
words removed. You may re-use your filters from Part A if you wish, although this is not required for full
credit.*

For the stop word list, I used the list of words originally returned by my Python output since all of the top 10 words returned could technically be defined as stop words. Hence, my population of stop words comprises: and, the, to, a, of, her, i, in, it, she. Below is the new top 25 list when these words are excluded.

In [8]:
!chmod +x stop.py

In [38]:
!cat women.txt |\
    ./split.py |\
    ./lower.py |\
    ./stop.py |\
    sort |\
    uniq -c |\
    sort -rn |\
    head -25 

   2250 you
   2224 for
   2034 was
   1977 as
   1853 with
   1848 that
   1503 he
   1463 but
   1247 jo
   1131 so
   1116 his
   1065 at
   1062 had
   1012 be
    974 on
    942 not
    916 if
    873 all
    843 my
    826 said
    819 is
    781 him
    749 me
    731 little
    715 have
sort: write failed: standard output: Broken pipe
sort: write error
